pkg/controller/elasticsearch/license/apply.go

```diff
@@ -23,42 +23,65 @@ import (
 var log = logf.Log.WithName("elasticsearch-controller")
 
 // isTrial returns true if an Elasticsearch license is of the trial type
-func isTrial(l *esclient.License) bool {
-	return l != nil && l.Type == string(esclient.ElasticsearchLicenseTypeTrial)
+func isTrial(l esclient.License) bool {
+	return l.Type == string(esclient.ElasticsearchLicenseTypeTrial)
+}
+
+// isBasic returns true if an Elasticsearch license is of the basic type
+func isBasic(l esclient.License) bool {
+	return l.Type == string(esclient.ElasticsearchLicenseTypeBasic)
 }
 
 func applyLinkedLicense(
 	ctx context.Context,
 	c k8s.Client,
 	esCluster types.NamespacedName,
-	current *esclient.License,
 	updater esclient.LicenseClient,
 ) error {
-	var license corev1.Secret
+	// get the current license
+	ctx, cancel := context.WithTimeout(ctx, esclient.DefaultReqTimeout)
+	defer cancel()
+	current, err := updater.GetLicense(ctx)
+	if err != nil {
+		return err
+	}
+
+	// get the expected license
 	// the underlying assumption here is that either a user or a
 	// license controller has created a cluster license in the
 	// namespace of this cluster following the cluster-license naming
 	// convention
-	err := c.Get(
+	var license corev1.Secret
+	err = c.Get(
 		types.NamespacedName{
 			Namespace: esCluster.Namespace,
 			Name:      esv1.LicenseSecretName(esCluster.Name),
 		},
 		&license,
 	)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			if isTrial(current) {
-				// this is the case if the user has started a trial on the ES cluster level
-				// e.g. from Kibana when trying to access a commercial feature.
-				// While this is not a supported use case we tolerate it to avoid a bad user
-				// experience because trials can only be started once on the ES cluster level
-				return nil
-			}
-			// no license linked to this cluster. Revert to basic.
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	if err != nil && apierrors.IsNotFound(err) {
+		// no license expected, let's look at the current cluster license
+		switch {
+		case isBasic(current):
+			// nothing to do
+			return nil
+		case isTrial(current):
+			// Elasticsearch reports a trial license, but there's no ECK enterprise trial requested.
+			// This can be the case if:
+			// - an ECK trial was started previously, then stopped (secret removed)
+			// - the user manually started a trial at the stack level (eg. by clicking a button in Kibana when
+			// trying to access a commercial feature). While this is not a supported use case,
+			// we tolerate it to avoid a bad user experience because trials can only be started once.
+			log.V(1).Info("Preserving existing stack-level trial license",
```

Comment:
Somehow 'preserving' sounded odd in my ears, but I have no better suggestion :-(